Add web app that follows the style of WinUI 3 and offers a design/layout similar to the Windows App

jackbuehner opened the pull request:
This pull request introduces a new PWA frontend (web app) for RAWeb that is in the style of WinUI 3 and Windows App. This PR also modernizes the default RemoteApp and desktop icon and adds support for icons created from wallpaper files. I have been slowly working on this since mid-March, so I hope I have added useful functionality and fixed most of the bugs!

## Install

Run as an administrator in PowerShell to install this branch:
```
$zipUrl = "https://github.com/jackbuehner/raweb/archive/refs/heads/winui.zip"; $tempDir = "$env:TEMP\raweb"; $zipFile = "$tempDir\master.zip"; New-Item -ItemType Directory -Path $tempDir -Force | Out-Null; Invoke-WebRequest -Uri $zipUrl -OutFile $zipFile; Expand-Archive -Path $zipFile -DestinationPath $tempDir -Force; Set-ExecutionPolicy Bypass -Scope Process -Force; & "$tempDir\raweb-winui\setup.ps1" -AcceptAll; Remove-Item -Path $zipFile -Force; Remove-Item -Path $tempDir -Recurse -Force;
```

## Limitations

- Firefox < 139 does not support `HTMLDialogElement.prototype.requestClose()` The properties dialog and terminal server picker dialog is disabled on Firefox and other browsers without support. Firefox 139 is in beta and is set to be released by the end of May. [Can I use...](https://caniuse.com/?search=HTMLDialogElement%20API%3A%20requestClose%20)
- Firefox and Safari do not support CSS Anchor Positioning and the `position-area` CSS property. Features that require context menus and dropdowns – including filters, viewing RDP file properties, and add and removing favorites – are disabled. Safari [recently added](https://developer.apple.com/documentation/safari-technology-preview-release-notes/stp-release-215?language=objc) CSS Anchoring Position support in TP 215, but as of TP 218, it lacks support for [implicit anchors](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_anchor_positioning/Using#implicit_anchor_association). [Can I use...](https://caniuse.com/mdn-css_properties_position-area)

## Changelog

### Additions

- WinUI-style frontend
  - New frontend following the style and layout of WinUI 3, offering a modern Windows-like experience
  - Login and logoff pages. This means we can now use autofill on the login page.
  - Web manifest for PWA support, including window controls overlay and install prompt
  - Service worker with caching similar to stale-while-revalidate. This enables faster loading and continued functionality in case the server with IIS is down.
  - Search the list of apps or devices
  - Ability to favorite apps and devices
  - Ability to import and export favorites
  - Ability to configure frontend-specific settings or force settings from appSettings in Web.config
  - Sort apps and devices by name, date modified, and terminal server
  - Filter apps and devices by terminal server
  - Card, tile, and list view modes
  - Simple mode, which is the closest equivalent to the current list of apps and desktops provided by RAWeb. It hides the sidebar and combines apps and desktops into a single page.

- Frontend installer (frontend/build.ps1)
  - Option 1: Copy source files directly to the RAWeb/app folder. Installation is faster, but page load times are slower.
  - Option 2: Create an optimized production build of the frontend using vite and copy the output to RAWeb/app. The build.ps1 script will download a copy of Node.js 22, install the required npm packages, and then build the frontend app.

### Changes

- The default.ico icon now matches Windows 11-style icons. It is the This PC icon, but the blue desktop background is tinted green to match RAWeb's green accent color.
- Desktop RDP files can now provide wallpaper as their icon instead of a square icon. When appropriate, the wallpaper will be converted into an icon that follows the style of the This PC icon from Windows 11, but the instead of the blue desktop background, it is the wallpaper.
- Dark mode versions of icons and wallpaper can be used in the RAWeb interface. Add -dark to the end of the icon name.
- setup.ps1 now also installs basic authentication. This is required for the webpage login interface.

### Fixes

- setup.ps1 no longer creates a virtual directory and an application for RAWeb. The virtual directory breaks the ability to use `ResolveUrl("~")`. Functionality appears to be the same with only the application in IIS.
- PNG images are resized to no larger than 256x256 before conversion to ICO. The function that converts PNG to ICO does not work with PNG icons larger that 256px in any dimension. This became apparent once the wallpaper icon functionality was added.

## Screenshots

| Light mode | Dark Mode |
| --- | --- |
| ![image](https://raw.githubusercontent.com/jackbuehner/raweb/59e225fa927fb4e72f1e1e91ceebe0b6345e99b2/frontend/lib/assets/favorites_light.png) | ![image](https://raw.githubusercontent.com/jackbuehner/raweb/59e225fa927fb4e72f1e1e91ceebe0b6345e99b2/frontend/lib/assets/favorites_dark.png) |
| ![image](https://raw.githubusercontent.com/jackbuehner/raweb/59e225fa927fb4e72f1e1e91ceebe0b6345e99b2/frontend/lib/assets/devices_light.png) | ![image](https://raw.githubusercontent.com/jackbuehner/raweb/59e225fa927fb4e72f1e1e91ceebe0b6345e99b2/frontend/lib/assets/devices_dark.png) |
| ![image](https://raw.githubusercontent.com/jackbuehner/raweb/59e225fa927fb4e72f1e1e91ceebe0b6345e99b2/frontend/lib/assets/apps_light.png) | ![image](https://raw.githubusercontent.com/jackbuehner/raweb/59e225fa927fb4e72f1e1e91ceebe0b6345e99b2/frontend/lib/assets/apps_dark.png) |
| ![image](https://raw.githubusercontent.com/jackbuehner/raweb/59e225fa927fb4e72f1e1e91ceebe0b6345e99b2/frontend/lib/assets/app-properties_light.png) | ![image](https://raw.githubusercontent.com/jackbuehner/raweb/59e225fa927fb4e72f1e1e91ceebe0b6345e99b2/frontend/lib/assets/app-properties_dark.png) |
| ![image](https://raw.githubusercontent.com/jackbuehner/raweb/59e225fa927fb4e72f1e1e91ceebe0b6345e99b2/frontend/lib/assets/login_light.png) | ![image](https://raw.githubusercontent.com/jackbuehner/raweb/59e225fa927fb4e72f1e1e91ceebe0b6345e99b2/frontend/lib/assets/login_dark.png) |
| ![image](https://raw.githubusercontent.com/jackbuehner/raweb/59e225fa927fb4e72f1e1e91ceebe0b6345e99b2/frontend/lib/assets/settings_light.png) | ![image](https://raw.githubusercontent.com/jackbuehner/raweb/59e225fa927fb4e72f1e1e91ceebe0b6345e99b2/frontend/lib/assets/settings_dark.png) |
| ![image](https://raw.githubusercontent.com/jackbuehner/raweb/59e225fa927fb4e72f1e1e91ceebe0b6345e99b2/frontend/lib/assets/terminal-server-picker_light.png) | ![image](https://raw.githubusercontent.com/jackbuehner/raweb/59e225fa927fb4e72f1e1e91ceebe0b6345e99b2/frontend/lib/assets/terminal-server-picker_dark.png) |

## Tested on

- [x] Chromium Desktop
- [x] Firefox Desktop
- [x] Safari Desktop
- [x] Chromim Mobile
- [x] Firefox Mobile
- [x] Safari Mobile

## Changes in detail

### .github/workflows/release.yaml

This new file creates a workflow that creates a new release for RAWeb. This workflow is manually triggered. It does the following:

1. Calculate a release version number. The format is vYYYY.MM.DD.0. This step checks if the latest release has the same year, month, and date. If so, it incriments the last number in the version number. For example, if the workflow runs on 2025-05-16, and the latest release is v2025.05.16.4, the calculated release version will be v2025.06.16.5.
2. Sets up nodejs and pnpm. It uses the same versions that are included at `/frontend/lib/bin/`. They are configured to cache.
3. Installs frontend dependencies with pnpm.
4. Builds the frontend app. The build process is configured to place all required files in `/apsx/wwwroot/`.
5. Moves `setup.ps1` to `/aspx/wwwroot/`. We do this so we can include it in the release archive.
6. Zips the `/aspx/wwwroot` folder and names it with the format `raweb_vYYYY.MM.DD.0.zip`.
7. Writes the `install.ps1` file, which includes the logic to download the zip file from the releases page and place everything in the required temporary folder structure such that `setup.ps1` can run.
8. Creates a draft release with install instruction and generated release notes (generated by GitHub from merged PRs).

To publish the release:

1. Go to the actions tab.
2. In the actions sidebar, click **Create draft release with prepared assets**.
3. At the top of the workflow runs table, click **Run workflow**.
4. In the menu, leave the branch on **master**. Click **Run workflow**.
5. Wait 20-60 seconds for the workflow to run.
6. When it is finished, go to the releases page on the repository.
7. On the new draft release that was just created, click the edit icon.
8. Make any required changes to the release. Do not change file names. If it is a pre-release, check the appropriate box. Click **Publish**. 

### aspx/wwwroot/Web.config

- Specifies mime types for `.vue`, `.mjs`, and `.webp` files.
- Disables caching for `.aspx`, `.css`, `.mjs`, `.rdp`, `.ts`, and `.vue`. The service worker from the frontend handles the caching now. Windows RADC and the other RD client apps did not refresh frequenctly enough for the cache to matter for them.

### aspx/wwwroot/App_Code/AliasUtilities.cs

This is the same utility in #37. It is now its own file so it can be imported and used in multiple places.

### aspx/wwwroot/Default.aspx

*Also: aspx/wwwroot/lib/controls/AppRoot.ascx*
*Removed: aspx/wwwroot/feedparser.js, aspx/wwwroot/lib/bootstrap.bundle.min.js, aspx/wwwroot/lib/bootstrap.min.css, aspx/wwwroot/lib/vue@3.js*

Now that the old frontend is replaced, `Default.aspx` is changed to work with the new frontend. It loads the script and stylesheet generated by the build step for the frontend. It references and loads `AppRoot.ascx`, which includes:

- `head` element, which loads the web manifest, the document favicon, and the css theme variable. It sets the title a theme color.
- The mount point, #app, for vue.
- A splash screen that includes the RAWeb logo and a loading icon. When the vuew app is ready, it hides the splash screen. The splash screen looks similar to the screens provided by many UWP apps.
- A short script that sets global variables on `window`. These variables include the path of the IIS application (e.g, /raweb/), the username of the authenticated user, the terminal server aliases from `Web.config`, and policies from `Web.config`.

### aspx/wwwroot/default.ico

Updated to use a modern icon. I extracted the This PC icon from Windows 11, adjusted the color of the desktop wallpaper portion from blue to green, and added the remote desktop icon (the two chevrons facing each other) as an overlay on top of the wallpaper portion.


aspx/wwwroot/bin/System.DirectoryServices.AccountManagement.dll, aspx/wwwroot/bin/System.DirectoryServices.dll


### aspx/wwwroot/get-image.aspx.cs

*Also: aspx/wwwroot/desktop-frame.png*

Icons can now have a dark variant. In the `resources` or `multiuser-resources` folder, a PNG icon with -dark at the end of the filename will be used (when available) when the `theme=dark` query string is provided in the URL. `get-image` will check if the dark variant file exists, and if it does not, it will use the light theme icon instead. The new functions are `FindImageFilePath` amd `ComposeDesktopIcon`.

Icons can be placed inside the central portion of `desktop-frame.png`. This is the frame from the This PC icon on Windows 11. When `frame=pc` is included in the query string, the requested image will be placed inside the desktop frame. 
This change required that the existing logic to serve ICO and PNG files was modified to operate based on `MemoryStream` instead of directly serving files. Now, the requested icon is read into a `MemoryStream` once it is determined to exist. Then, if it should be placed into a desktop frame, a new `MemoryStream` is created that includes the image inside the desktop frame. Then, the `MemoryStream` in served to the client in the requested format.

When converting to ICO format before serving the image as ICO, the `MemoryStream` image is resized such that it never exceed a width or height of 256 pixels. The current implimentation of PNG to ICO does not work with images larger than this size.


### aspx/wwwroot/login.aspx, aspx/wwwroot/login.aspx.cs

The login pages renders a form that allows login with credentials. This is an improvement over /auth/login, which shows a browser popup that does not support autofill by password managers.

Here is how the login flow works:

1. User provides credentials and clicks the submit button.
2. `login.aspx.cs` checks if the username and password are valid for the domain. It is checked with `PrincipalContext.ValidateCredentials`, which is case sensitive for usernames and passwords.
  - If they are invalid, an error message appears on the login page.
  - If they are valid, it injects an invisible button that triggers the login from the browser, and then it clicks the button
3. The `authenticateUser` function is triggered in `login.aspx`. This function is client-side, and it submits a POST request to `/auth/login.aspx` with the credentials. If the response is okay, it redirects to the frontend. If `ReturnUrl` is in the search params, it redirects to that URL instead (e.g., the user was redirected to login after trying to load /Default.aspx#settings, so they should be redirected back to that page).

The reason we do it this way is because the browser login popup will appear if we submit a POST request with invalid credentials. By checking the validity of the credentials first, we can ensure that the login popup does not appear.


### aspx/wwwroot/logoff.aspx

This page clears the session cookie and authentication cookie. It also clears cached data from localStorage and wipes the service worker cache. WQhen finished, it redirects to the login page.


### aspx/wwwroot/lib/controls/Header.ascx

Provides a titlebar for the login and logoff pages that works with window controls overlay


### aspx/wwwroot/lib/controls/InfoBarCritical.ascx

Provides a control for rendering a critical-stlye info bar. This is used on the login page for showing errors.


### aspx/wwwroot/lib/controls/head.ascx

Provides a control for setting the window title and loading assets such as the web manifest, favicon, and css variables stylesheet. It is used on the login and logoff pages.


### aspx/wwwroot/webfeed.aspx

- Desktops now place their icons inside the desktop frame. `frame=pc` is now included as part of the icon URL for the icon elements. If a wallpaper is not available, the new default wallpaper is used instead. Wallapers should be named <rdp_file_name>.png. Unlike other icons, wallpapers can be any dimension. They will be a square when provided to clients because they are inserted into the square desktop-frame.png icon.
- GetResourceGUID now sorts the lines alphabetically. This ensures that RDP files that are duplicates aside from the order of the lines are now properly marked as duplicates.
- A new option to merge terminal servers allows combining resources that are the same aside from the terminal server address (full address). These resources will included both terminal servers in their `HostingTerminalServer` element. When this option is enabled, GetResourceGUID omits the `full address` line from the RDP file, allowing the RDP files that are identifcal aside from the terminal server to have the same GUID. This option is enabled by providing the `mergeTerminalServers=1` in the query string. This option must be explicitly provided because it does not work will all RD clients. The RAWeb frontend supports this option and can be configured to use it by default via the settings page or `Web.config`.
- `AliasResolver` is moved to `aspx/wwwroot/App_Code/AliasUtilities.cs`.


### frontend/bin/

This folder includes binaries for nodejs and pnpm. These can be used to build the app during development. Most users will use the generated release, which contains the already-built frontend. The generated release only includes the aspx/wwwroot folder, so the download size will not be affected by the presence of these binaries.


### frontend/build.ps1

This script can be called by `setup.ps1` when the frontend is not already built.


### frontend/lib/App.vue

This is the main part of the frontend. It handles routing; loading the service worker; hiding the splashs screen; applying view transitions; tracking prefers-reduced-motion; and rendering the titlebar, navigation bar, and pages.


### frontend/lib/public/manifest.aspx

This file creates the web manifest for the RAWeb frontend. This is what makes it possible to manually install RAWeb's frontend as an app. The manifest declares `window-controls-overlay` as a supproted display mode, which makes it possible to merge the OS and RAWeb titlebar into a single titlebar.


### frontend/lib/public/service-worker.js

This file manages the caching for the frontend. All files in `/resources/`, `/multiuser-resources/`, and `/lib/` are cached by the service worker. `get.image.aspx`, `Default.aspx`, and `icon.svg` are also cached. The caching emulates `stale-while-revalidate` caching; whenever a resource is requests, the cached version is immediately returned, and then the cached version is updated in the background.

If the service worker detects the `/auth/loginfeed.aspx`, it redirects to `/logoff.aspx`. The redirect to `/auth/loginfeed.aspx` only occurs when the user's credentials have expired or have been revoked, and then they try to fetch the webfeed.

The service worker stores a list of resources to update in the background (a "fetch queue"), and then it does not starting updating them in the background until all of the cached versions are used. This ensures that the page loads fast. When no more resources are being requested, the fetch queue is processed. The frontend's titlebar shows a loading spinner whenever the fetch queue is not empty.

When a service worker exists, chromium-based browsers will prompt users to install the frontend as an app.


### frontend/lib/types.d.ts

This file adds types for the variables that we inject into `window` in `AppRoot.ascx`. This ensures that typescript does not show errors when working on the vue app and trying to use those variables.


### frontend/lib/assets/

This folder contains the screenshots and icons for the web manifest. This enables the appropriate icon to appear when adding RAWeb as an app. The screenshots appear in the prompt to install (controlled by the browser).

This folder also includes the default wallpapers for desktops that do not specify their own wallpaper in the `/resources` or `/multiuser-reosurces` folders.

The `icons.ts` file includes the icons used by the frontend. These are the `svg` icons that appear in the sidebar, titlebar, and menus.


### frontend/lib/components/

This folder contains all of the re-usable components used by the frontend. Each component is re-exported at `frontend/lib/components/index.mjs`.


### frontend/lib/pages/

This folder contains the pages for the frontend. ROuting is handled in `frontend/lib/App.vue`, but each route loads its associated page component from this folder.


### frontend/lib/utils/PreventableEvent.ts

This file is a constructor for a custom event used in the `ContentDialog.vue` component. This event sets `defaultPrevented` to `true` when `evt.preventDefault()` is called, and then `ContentDialog` will not act on the event.

`TerminalServerPickerDialog.vue` calls `evt.preventDefault()` so that it can perform some checks before actually closing the dialog.


### frontend/lib/entry.dist.mjs

This is the entrypoint for the frontend. This file is in charge of creating the vue app from `frontend/lib/App.vue` and moun ting it to `#app` in `Default.aspx`. Vite starts crawling for assets related to the frontend


### frontend/lib/utils/combineTerminalServersMode.ts

This tracks whether the setting for comining terminal servers is enabled. If `combineTerminalServersModeEnabled` is set in `Web.config`, it uses the value from there. Otherwise, it stores the value in `localStorage` with key `combine-terminal-servers-mode:enabled` and restores it on page load. When enabled, the `mergeTerminalServers=1` query is set when requesting `webfeed.aspx`.


### frontend/lib/utils/flatMode.ts

This tracks whether the setting for flat mode is enabled. If `flatModeEnabled` is set in `Web.config`, it uses the value from there. Otherwise, it stores the value in `localStorage` with key `flat-mode:enabled` and restores it on page load. Flat mode hides all folders on the apps page.


### frontend/lib/utils/getAppsAndDevices.ts

This fetches and processes the MS-TSWP version 2.0 webfeed.

RDP files are downloaded and coverted into a object of keys and values that correspond to each line in the RDP file. The object is stored for each unique resource for each terminal server. For example, an RDP file with:

```text
remoteapplicationmode:i:1
remoteapplicationprogram:s:C:\Windows\explorer.exe
```

becomes

```js
{
  remoteapplicationmode: 1,
  remoteapplicationprogram: 'C:\Windows\explorer.exe',
}
```


### frontend/lib/utils/generateRdpFileContents.ts

This file converts an object containing all RDP file properties into a text file that can be downloaded as an RDP file.

The object representation of the RDP file generated by `getAppsAndDevices.ts` is converted back to an RDP file with this function when the user requests to download the RDP file.

Why is it implimted this way? This allows us to render the RDP file contents in a nice properties dialog for each resource. This dialog can be accessed by right click any resource in the frontend. In the future, we may want also want to use this to add a setting to insert a username and password for resources from a specific terminal server.


### frontend/lib/utils/iconBackgrounds.ts

This tracks whether the setting for showing icon backgrounds is enabled. If `iconBackgroundsEnabled` is set in `Web.config`, it uses the value from there. Otherwise, it stores the value in `localStorage` with key `icon-backgrounds:enabled` and restores it on page load. When icon backgrounds are enabled, resource icons are shown an a solid white (or dark gray when in dark mode) background when the list of resources is set to card mode. This is similar to the appearance of Windows App on Windows.


### frontend/lib/utils/simpleMode.ts

This tracks whether the setting for simple mode is enabled. If `simpleModeEnabled` is set in `Web.config`, it uses the value from there. Otherwise, it stores the value in `localStorage` with key `simple-mode:enabled` and restores it on page load. When simple mode is enabled, the sidebar is hidden, and the favorites, apps, and devices pages are all replaced with a single page that shows all apps and devices in a single list. The list has folders support. This is most similar to the current RAWeb frontend.


### frontend/lib/utils/unproxify.ts

Vuew stores many objects in refs as proxies. This is a helper function to recurses through the proxy until it finds the real value, and then it returns it.


### frontend/lib/utils/useFavoriteResources.ts

This tracks which resources (apps or desktops) are favorited by the current logged in user. They are stored in `localStorage` as an array of resources in key `favorite-resources`.

This also tracks whether the setting for showing favorites is enabled. If `favoritesEnabled` is set in `Web.config`, it uses the value from there. Otherwise, it stores the value in `localStorage` with key `favorite-resources:enabled` and restores it on page load. When favorites are disabled, favorites is hidden from the sidebar navigation, and the option to add and remove favorites is hidden. When disabled, the favorites route redirects to the apps route.


### frontend/lib/utils/useWebfeedData.ts

This is a helper for calling `getAppsAndDevices`. It creates reactive variables for the data or errors returned by `getAppsAndDevices`, a reactive variable for the loading state, and it provides a `refresh` function for re-fetching the webfeed data and updating the reactive variables. This is preferred way to get the webfeed data in the app. Components should import and use the `useWebfeedData` function.


### frontend/lib/winui.css

This is the file that contains all of the theme variables (colors, font sizes, etc.) for the frontend. Colors were extracted from the theme resources specified in WinUI 3 Gallery. Dimensions and font sizes also came from WinUI 3 Gallery.


### frontend/package.json

This file includes the dependencies for the frontend.

It also includes the scripts for developing or building the frontend.

To develop (watch files for changes and automatically rebuild): `pnpm run dev`

To build and optimize for production: `pnpm run build` OR run `.frontend/build.ps1`

If nodejs and pnpm are not installed, run `./frontend/build.ps1` to temporarily add `/frontend/bin/` to your path. You will need to do this every time you open a new powershell window.

Do not delete `frontend/pnpm-lock.yaml`; it includes the specific package versions that were resolved from `package.json`.


### frontend/tsconfig.json

This file configures typescript. It sets which files should be seen by the type checker, and it also specifies how to resolve the import aliases (e.g., $components).


### frontend/vite.config.ts

This file contains the vite configuration for the build process.

- The vue plugin adds support for vue single-file components.
- The custom plugin called 'raweb:copy-assets` copies all assets needed by the app to `/aspx/wwwroot/`.
- The `resolve` option tells vite how to resolve import aliases.
- `base: './'` tells vite that we do not know the axact path where the app will be deployed, so it should use relative imports. There is no guarantee that the frontend is deployed to /raweb/.
- The `build` option configures vite to output the bundled assets to '/aspx/wwwroot/'. We configure the `build.rollupOptions.input.main` option because the default is index.html, but we use `/frontend/lib/entry.dist.mjs`. The `rollupOptions.output` options ensure that random letters and numbers are not included at the end of the filename. We need the file names to be consistent because they are imported by `Default.aspx`.


### setup.ps1

- Basic authentication is also enabled. Basic authentication is required for the login page. Basic authentication is now downloaded and enabled by the setup script.
- A new AcceptAll switch allows for non-interactive installs. It accepts all prompts as 'Y'. Example `./setup.ps1 -AcceptAll`.
- The debug information can now be shown with the Debug parameter instead of needing to manually set `$debug = $true`. Run with the `-Debug` switch to view the debug messages.
- The script no longer creates a virtual directory and an application for RAWeb. The virtual directory breaks the ability to use ResolveUrl("~"). Functionality appears to be the same with only the application in IIS.
- If the script cannot find the built version of the frontend, it will attempt to run '/frontend/build.ps1`. This is only needed when attempting to install RAWeb by downloading or cloning the repository and then running `setup.ps1`.
- Files in the "resources" or "multiuser-resources" are no longer deleted or overwritten when re-installing RAWeb.

jackbuehner commented:
## To-do

- [ ] Test on Safari
- [ ] Test on Firefox
- [ ] Write documentation

kimmknight commented:
> @kimmknight This currently creates a new directory at /raweb/app. Do you want to leave the existing Default.aspx (/raweb) in addition to /raweb/wapp, or should this completely replace it?

I reckon completely replace it.

jackbuehner commented:
> Server Error in '/RAWeb' Application.

I messed up the merge 🤦. It caused that code that block of code to appear an extra time, before the correct line where `injectionPointElement` is defined. I just pushed a fix.



> > @kimmknight This currently creates a new directory at /raweb/app. Do you want to leave the existing Default.aspx (/raweb) in addition to /raweb/wapp, or should this completely replace it?
> 
> I reckon completely replace it.

Sounds good. I can work on making that happen.


> As for the setup, I'm trying to keep it as simple and idiot-proof as possible. In theory (without testing), my preference would be to not present the choice of whether to download npm and bundle, and instead, just deploy the unbundled version. How big of a performance difference are we talking?

I agree, that's generally my preference too (and the reason the unbundled version exists). The issue is that most browsers will only allow a few concurrent network requests per origin. The vue app is now using single-file components because it would be too unweildy to keep the whole app in only a few files, there are a lot more files to download in order to load the app. The service worker cache helps a bit with this after the first load.
On my local network, the bundled version loads in just under 0.7s, but the unbundled version takes around 2.7s. If I switch I load the app from a different location, the bundled version loads also loads in less than a second, but the unbundled version takes around 8 seconds.

When I think about it more, having two methods is only going to cause issues in the future. It may be most ideal to set up a GitHub Actions workflow that builds the vue app and creates a new release with a ready-to-use .zip containing the combined contents of the aspx folder and the bundled assets whenever a new tag is pushed to the repository. The setup script can then be updated to download from the latest release. What do you think about that option?

kimmknight commented:
> On my local network, the bundled version loads in just under 0.7s, but the unbundled version takes around 2.7s. If I switch I load the app from a different location, the bundled version loads also loads in less than a second, but the unbundled version takes around 8 seconds.

Thanks - great to have some stats. Yeah, 8 seconds is significant. Not ideal, but probably acceptable for most RAWeb users. If there were no workaround, I'd suggest just leave it unbundled, however, your idea below sounds good - if it's reasonably achievable and maintainable.

> When I think about it more, having two methods is only going to cause issues in the future. It may be most ideal to set up a GitHub Actions workflow that builds the vue app and creates a new release with a ready-to-use .zip containing the combined contents of the aspx folder and the bundled assets whenever a new tag is pushed to the repository. The setup script can then be updated to download from the latest release. What do you think about that option?

Sounds like a great idea. I am inexperienced with GitHub Actions. Would you like me to add you as a repo collaborator?

jackbuehner commented:
I'm not 100% sure how GitHub actions will respond to me commiting workflows in a PR to a repo I do not collaborate on, so adding me may be a good idea. Either way, I'll try to test the workflow on my fork so I do not pollute the actions history of the main repository while testing.

kimmknight commented:
I tested it again tonight - working 👌

This UI is sooooo cool!

jackbuehner commented:
An example of the action and generated release can be seen at:
- Action: https://github.com/jackbuehner/raweb/actions/runs/15057761368/job/42327087216
- Release: https://github.com/jackbuehner/raweb/releases/tag/v2025.05.16.0

kimmknight commented:
> > Just discovered that username is case sensitive
> 
> I honestly had no idea that windows authentication is not usually case sensitive. The login page [uses](https://github.com/kimmknight/raweb/pull/44/files#diff-e16efe3208f1bd6f1e07f1befd6968ac4103bfe46102466415035350aad6436aR67) [PricipalContext.ValidateCredentials](https://learn.microsoft.com/en-us/dotnet/api/system.directoryservices.accountmanagement.principalcontext.validatecredentials?view=netframework-4.5) to sign in with Windows credentials, and it is case sensitive. Do we need to find a workaround to allow case insensitivity?

Yeah, Windows accounts are largely case-insensitive within the OS and AD. It looks like PrincipalContext.ValidateCredentials is case-sensitive depending on the identity provider. Apparently it is unlikely to be case-sensitive for AD accounts, however, it (strangely) treats Local accounts case-sensitively. Not a show-stopper, but maybe (in future) we mention it on the login page to avoid confusion for typical Windows users.

kimmknight commented:
> An example of the action and generated release can be seen at:
> 
> * Action: https://github.com/jackbuehner/raweb/actions/runs/15057761368/job/42327087216
> * Release: https://github.com/jackbuehner/raweb/releases/tag/v2025.05.16.0

So awesome! I'll give it a good test later tonight 👍

jackbuehner commented:
I did a bit more investigation into local accounts. It turns out that the logic `/login.aspx.cs` is _not_ case sensitive. Instead, `/auth/login.aspx` is where the case sensitivity issue is occuring. Perhaps basic authentication for IIS is case sensitive.

I will add a step in `/login.aspx.cs` that will get the correct case username.

jackbuehner commented:
Alright - I think it's ready to go. I have tested on Edge/Chrome, Firefox, and Safari. I have also updated the main PR description with some limitations on Firefox and Safari and a _changes in detail_ section.

kimmknight commented:
This is incredible! Do we need to add and run the action before merging this? Also, README?

kimmknight commented:
With this new UI and features, I think the repo description kind of undersells it:

`A simple web interface for your RemoteApps hosted on Windows 10, 11 and Server.`

jackbuehner commented:
> This is incredible! Do we need to add and run the action before merging this? Also, README?

Glad you like it!

The action needs to run after merging. It needs to be manually triggered whenever a new version is ready to be released. Once this is merged, the release can be created with the following steps:

1. Go to the actions tab.
2. In the actions sidebar, click Create draft release with prepared assets.
3. At the top of the workflow runs table, click Run workflow.
4. In the menu, leave the branch on master. Click Run workflow.
5. Wait 20-60 seconds for the workflow to run.
6. When it is finished, go to the releases page on the repository.
7. On the new draft release that was just created, click the edit icon.
8. Make any required changes to the release notes. Do not change file names. If it is a pre-release, check the appropriate box. Click Publish.

I just made an attempt at an updated README. It includes the updated screenshots, an expanded features list, and the new installation method. The old installation methods have been collapsed under the _Other installation methods_ section. I also moved the _Publishing RemoteApps and Desktops_ section to a wiki page and replaced it with a list of helpful resources from the wiki. Feel free to make changes.

eangulus commented:
I followed the instructions and It went thru without errors in powershell.

But when I go to the site I now see:

`

Server Error in '/' Application.
Parser Error
Description: An error occurred during the parsing of a resource required to service this request. Please review the following specific parse error details and modify your source file appropriately.

Parser Error Message: The file '/lib/controls/AppRoot.ascx' does not exist.

Source Error:

Line 1: <%-- This page should only be used in the version of the app built with vite. --%> Line 2: <%@ Page Language="C#" AutoEventWireup="true" %> Line 3: <%@ Register Src="~/lib/controls/AppRoot.ascx" TagName="AppRoot" TagPrefix="raweb" %> Line 4: Line 5: --

Source File: /RAWeb/default.aspx    Line: 3


Version Information: Microsoft .NET Framework Version:4.0.30319; ASP.NET Version:4.8.9282.0

jackbuehner commented:
Is the C://ine pub/raweb/lib/controls/AppRoot.ascx file there when you look for it with file explorer?

If you remove the RAWeb application from IIS and re-run setup.ps1, does it work? This will not delete your resources or multi-user resources. The old setup script used to create a virtual directory in addition to the application, but this causes some issues. If removing from IIS and then re-running setup.ps1 works, I'll be able to write a fix.

jackbuehner commented:
@eangulus, thank you for sharing the problem you had. I think I was able to replicate the issue, and I just pushed a fix.

The setup.ps1 script no longer creates an IIS virtual directory because it causes issues with resolving certain paths (which is what you experienced). I neglected to leave the part of the script that removed the virtual directory when running setup.ps1 on a server that already has RAWeb. The PR adds that step back, so this should not be an issue again.

eangulus commented:
No problem. I'll test again later and let you know how I went.

eangulus commented:
Where is the updated version?

Can I download? Or do I just run the script for installing again?

jackbuehner commented:
> Where is the updated version?
> 
> Can I download? Or do I just run the script for installing again?

Running the script again should do it. It automatically downloads https://github.com/jackbuehner/raweb/archive/refs/heads/winui.zip, which is the latest version of this branch.

eangulus commented:
I just did this and I don't get the same error as before, but I also cannot login.

Mine is a domain joined windows 11 VM, I get the in page html login page, when I try to login using a variety of users and also with DOMAIN\User format, I then get the popup login like in the original version of this. I try same username and password there and it just constantly fails piping up a new login popup again and again.

jackbuehner commented:
> I just did this and I don't get the same error as before, but I also cannot login.
> 
> Mine is a domain joined windows 11 VM, I get the in page html login page, when I try to login using a variety of users and also with DOMAIN\User format, I then get the popup login like in the original version of this. I try same username and password there and it just constantly fails piping up a new login popup again and again.

I set up a new domain controller and did some testing. The login page did not fully support domain accounts before, but it does now (https://github.com/kimmknight/raweb/pull/44/commits/509fb8308851cd20f1a2ca68004adaa68059ae82). Sorry about that!

If it still does not work, you may need to go into IIS manager and configure the default domain for basic authentication. ([Instructions](https://www.omnisecu.com/security/internet-information-services-iis-7-security/how-configure-basic-authentication-in-iis-7.php)) I don't have enough system resources to set up another VM to see if this is necessary for a domain-joined computer/server that is not a domain controller. Let me know if you end up needing to do this - we might be able to detect whether a computer is domain-joined in `setup.ps1` and then configure the default domain as part of setup.

I'm not really sure why the popup didn't work either (it worked from /raweb/auth/login.aspx), but it does not matter since the login popup should not appear when the authentication process is working correctly.

eangulus commented:
Will attempt again soon. Thanks for the quick responses.

eangulus commented:
Confirmed working now.

I didn't have to do anything but install via the script.

eangulus commented:
Corretion.

Small issue. Not 100% sure if its my end or not, but I have tried everything so far.

So, I currently have 2 Users that can use my system. A Win11 Domain Joined PC. BOth users are in a security group, and I have allow this security group to have remote access in the Remote Desktop Settings.

Both users can login directly to the VM. Both have access to the apps etc.

Both users can use RemoteApp and desktop connections setup using the webfeed URL.

But for some reason, Only 1 user can login using the RAWeb site. The other user doesnt show any errors or failed login info or anything like that, it just refreshes to the login again.

I have cleared cache and I have tried multiple computers (just in case cache was at play) and nothing. Still only the 1 out of otherwise 2 identical user accounts can login to the RAweb page.
